### PR TITLE
Add tutorial for preprocessing API

### PR DESCRIPTION
**Pull request by sbalandi**

### CODE REVIEW

**Comment by eaidova** [comment]
@sbalandi please include notebook to README.md
https://github.com/openvinotoolkit/openvino_notebooks/blob/main/README.md
https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/README.md

File: notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb [hunk 1]
```diff
@@ -0,0 +1,498 @@
+{
```
Comment (eaidova) [inline notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb:1]: <p>Can we do not use OMZ here?</p><p>Just download model from original source and convert it using mo</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/811/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (eaidova) [inline notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb:1]: <p>I think, it is better use python style of class naming for avoid misleading.</p><p> </p><p>Also I think read from disk is not suitable clarification here, can we remove it?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/811/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (eaidova) [inline notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb:1]: <p>from what this input shape come?</p><p>It also will be useful if you can demostrate this info you using image for better explanation how this was obtained (e.g. read input image and print its shape, and data type ) </p><p> </p><p>Maybe it is also be useful to specify how to fit input data if input shape can be dynamic (e.g. what if we want process several images with different shapes)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/811/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (eaidova) [inline notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #5.                             .mean([104.0, 117.0, 123.0])
</pre><p>can be these commented steps described as part of markdown?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/811/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (eaidova) [inline notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #14.    resized_image = np.array([104, 117, 123]).reshape((3, 1, 1))
</pre><p>it is not resized image, it is mean_values. Probably order should be slightly different (resized_image - mean_values, then create input_tensor, it helps to avoid additional reshaping) and missed step for converting input data type from uint8 to float32</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/811/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

File: notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+    "\n",
+    "end = time.perf_counter()\n",
+    "time_ir = end - start\n",
+    "\n",
+    "print(\n",
+    "    f\"IR model in OpenVINO Runtime/CPU with preprocessing API: {time_ir/num_images:.4f} \"\n",
+    "    f\"seconds per image, FPS: {num_images/time_ir:.2f}\"\n",
+    ")\n",
+    "\n",
+    "# check performance for inference with the manual preparing of the image\n",
```
Comment (OpenVINO-dev-contest) [inline notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb:450]: Shall we add opencv manual processing into this loop, which is also included in execution graph with preprocessing API

**Comment by sbalandi** [comment]
> @sbalandi please include notebook to README.md https://github.com/openvinotoolkit/openvino_notebooks/blob/main/README.md https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/README.md

done

File: notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb (hunk 1 above)
Comment (sbalandi) [reply notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb:1]: yes, done

Comment (sbalandi) [reply notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb:1]: done

Comment (sbalandi) [reply notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb:1]: moved

Comment (sbalandi) [reply notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb:1]: yes, thank you, changed

File: notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb (hunk 2 above)
Comment (sbalandi) [reply notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb:487]: done

File: notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb (hunk 1 above)
Comment (sbalandi) [reply notebooks/117-optimize-preprocessing/117-optimize-preprocessing.ipynb:1]: _from what this input shape come?_ - what do you mean ? where I found the values ?
_It also will be useful if you can demostrate this info you using image for better explanation how this was obtained (e.g. read input image and print its shape, and data type )_ - done
_Maybe it is also be useful to specify how to fit input data if input shape can be dynamic (e.g. what if we want process several images with different shapes)_ - fitting image is described below in section **Load image and fit it to model input**, also a little bit about dynamic input shapes there is in **Preprocessing Steps**

**Comment by eaidova** [comment]
@sbalandi 117 used for model-server, please move to 118

File: notebooks/118-optimize-preprocessing/118-optimize-preprocessing.ipynb [hunk 3]
```diff
@@ -0,0 +1,537 @@
+{
```
Comment (eaidova) [inline notebooks/118-optimize-preprocessing/118-optimize-preprocessing.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #27.        mo_command = f"""mo
</pre><p>model = mo.convert_model(...)</p><p>BTW, looks like you subtract mean_values twice, this parameter should be removed. It is equivalent of mean_values in preprocessing api</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/811/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (eaidova) [inline notebooks/118-optimize-preprocessing/118-optimize-preprocessing.ipynb:1]: <p>peak of popularity of caffe was in 2018, for now it os almost dead framework. Maybe we can consider different model example?</p><p>e.g. some  model from here https://www.tensorflow.org/api_docs/python/tf/keras/applications</p><p>Additionally it contains preprocessing function, which we can use as reference</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/811/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (eaidova) [inline notebooks/118-optimize-preprocessing/118-optimize-preprocessing.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #6.    if Path(MODEL_DIR, f'{model_name}.prototxt').exists() and Path(MODEL_DIR, f'{model_name}.caffemodel').exists():
</pre><p>this check unneccesary, download_file writes the same message if file exists</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/811/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (eaidova) [inline notebooks/118-optimize-preprocessing/118-optimize-preprocessing.ipynb:1]: <p>&gt;&gt;Therefore, in this case, target height / width should be specified as PreProcessSteps.resize( ResizeAlgorithm, destination_height, destination_width).</p><p> </p><p>this sentence is not clear for me, as I know there is no opportunity to specify target height/width in preprocessing API. it uses model input shape instead. e.g. if model has input shape [1, 3, 224, 224], resize will use 224x224 as output size.</p><p> </p><p>I mean different when speak about dynamic.</p><p> </p><p>set_shape([1, 577, 800, 3]) - this method declare that all processed data always  have this size. It is good for example, when you process video, because all frames has the same resolution, right?</p><p>But we can not say it about fitting just images. If you want process image with different size, it should be resized to 800x577 before submitting to model or preprocessing should be reconfigured and rebuild again for next resolution and it is for each image, which can be not convinient.</p><p> </p><p>There is another way as I know to describe tensor for this condition. </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/811/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (eaidova) [reply notebooks/118-optimize-preprocessing/118-optimize-preprocessing.ipynb:1]: I think we also should add some imention about mean/scale values and reverse_input_channels using mo as alternative and avoiding double work like in your case

Comment (eaidova) [reply notebooks/118-optimize-preprocessing/118-optimize-preprocessing.ipynb:1]: here example how it can be used with openvino:
https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/101-tensorflow-to-openvino/101-tensorflow-to-openvino.ipynb

Comment (eaidova) [reply notebooks/118-optimize-preprocessing/118-optimize-preprocessing.ipynb:1]: I'm not sure, is there mention about it in docs,
set_spatial_dynamic_shape() used for cases where the original image size is unknown and can be any instead of tensor set_shape

Comment (sbalandi) [reply notebooks/118-optimize-preprocessing/118-optimize-preprocessing.ipynb:1]: add description and change to mo.convert_model

Comment (sbalandi) [reply notebooks/118-optimize-preprocessing/118-optimize-preprocessing.ipynb:1]: changed to InceptionResNetV2

Comment (sbalandi) [reply notebooks/118-optimize-preprocessing/118-optimize-preprocessing.ipynb:1]: thank you ! yes, it is not mention in overview or details, I found it late in https://docs.openvino.ai/latest/openvino_docs_OV_UG_Preprocess_Usecase_save.html .

File: notebooks/118-optimize-preprocessing/118-optimize-preprocessing.ipynb [hunk 4]
```diff
@@ -0,0 +1,533 @@
...
```
Comment (eaidova) [inline notebooks/118-optimize-preprocessing/118-optimize-preprocessing.ipynb:1]: <p>can we split this into 2 parts:</p><ul><li>convert model with preprocessing using mo</li><li>convert model without preprocessing in mo and add preprocessing using preprocessing API</li></ul><p>and then compare these 2 steps (and see that these steps lead to the sample performance/accuracy of results)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/811/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (eaidova) [reply notebooks/118-optimize-preprocessing/118-optimize-preprocessing.ipynb:1]: tensorflow preprocessing is fine :)

**Review by eaidova — APPROVED** [review]
